\documentclass{llncs}
\usepackage{graphicx}
%\usepackage{natbib}
%
\begin{document}

\title{Preparing KEGG pathway models for various applications}
\titlerunning{KEGG pathways as models}

\author{Clemens Wrzodek\inst{1} \and Andreas Dr\"ager\inst{1} \and Finja B\"uchel\inst{1} \and Manuel Ruff\inst{1} \and Andreas Zell\inst{1}}
\authorrunning{Wrzodek \emph{et al.}}

\institute{Center for Bioinformatics Tuebingen (ZBIT), University of Tuebingen, T\"ubingen, Germany}

\maketitle


\begin{abstract}
%problem statement:
%      what problem are you going to solve?
%
%motivation/relevance:
%      why is it important to solve this problem?
%
%approach/method:
%      how did you go about solving the problem?
%
%results:
%      what is your solution to the problem?
%
%conclusions:
%      what are the implications of your solution?
\end{abstract}

\section{Introduction}
%Opening statement
%    Problem: what problem are you going to solve?
%    Relevance: why is it important to solve this problem?
%
%Literature review:
%    Summarize the current state of knowledge in the area of
%              investigation
%    What previous research has been done on the problem?
%    Mark the gap in the current state of knowledge:
%    What piece of knowledge is still lacking in the area of
%              investigation?
%
%Brief overview of your own study:
%    Approach/method: how did you go about filling this gap?
%    Results: what did you find out?
%    Conclusion: what do these results mean (why are these
%               results significant?

% 1. KEGG ist toll weil
% 2. Modelling
% 3. ...und man könnte tolle Sachen machen wenn man tolle Modelle hätte.
% 4. Deshalb gibt es Converter
The KEGG PATHWAY database provides a valuable resource for initial modeling approaches of specific biological networks \cite{Kanehisa2000}. The database contains pathway maps for a multitude of different organisms and most provided information is cross-linked with other KEGG databases. Since many years, this database is one of the most important sources for building initial structural models of various pathways \cite{Bauer-Mehren2009,Oberhardt2009}. All pathway information is stored in KGML formatted xml-files, which are barely supported by other applications. In systems biology, two wide-spread formats for modeling and exchanging pathways are the SBML and BioPAX formats (TODO: Cite SBML and BioPAX). These formats can be used with graphical modeling applications, like CellDesigner (), annotated with kinetics (squeezer subliminal), used for flux balance analysis (fasimu), etc. (TODO anwendungen für BioPAX finden). Therefore, converters exist that perform mostly basic conversions from KGML to those formats \cite{KEGGconverter,KEGG2SBML}. The drawback of many of those conversions is, that even for creating initial models a basic translation of a KGML document to, e.g., an SBML document is not sufficient.

% 5. Wo ist das Problem mit den bisherigen Convertern?
The KGML documents provided by KEGG are mainly for graphical representations of pathways and consist of entries (which correspond to nodes in a pathway map), relations (which correspond to edges in a pathway map) and reactions. Those reactions are primarily contained in metabolic pathway maps and allow to draw the conclusion that converting all entries and reactions would already lead to good metabolic models. But a closer look on the actual maps shows that even those reactions are mainly for visualization and not for modeling or simulation purposes. Reactions are sometimes being "bundled", e.g., one reaction instance is built and multiple reaction identifiers, that point to different reactions, are assigned. There are often missing reactants for reactions, stoichiometric information is completely wiped out and also the list of enzymes, catalyzing a reaction, is not necessarily entirely contained in the KGML document. Similar problems arise for the entries in a KGML document. The pathways contain some duplicates of entries, which make sense for nice graphical representations but not for modeling and simulation. Furthermore, one entry might consist of multiple proteins/ genes, etc. which are grouped together for some reason. All those exemplary mentioned problems show that simple one-to-one translations of KEGG pathway maps are not sufficient to build reliable and useful models.

% 6. Wie sind besser und lösen alles.
To overcome all those drawbacks, we performed many case-studies on those KGML documents and developed strategies for building useful initial models in SBML and BioPAX. Besides automatically fixing many of the mentioned issues, we heavily annotate and augment all provided information to ease further model building and usage of those translated pathway maps. This ranges from adding simple database cross-references (e.g., to UniProt or Entrez Gene), over annotation of chemical formulas and molecular weight of small molecules, to an automated atom balance check of all reactions. All those strategies are now included in the second release of the KEGGtranslator application \cite{Wrzodek2011}.



% 1. KEGG ist toll weil
% 2. Wo ist das Problem (beschreibt grafische Zusammenhänge)
% Verschiedene Formate für versch. Sachen, nicht verlustfrei möglich
% Man braucht (korrekte) Gleichungen, am besten noch chemische formeln, molekular gewicht, etc.
% für graphen malen braucht man größe, position, farbe, noch entrez gene, uniprot (und kanten)
% Inhalt ist für grafik, Reaktionen da fehlt einiges, sowie annotationen, etc.
% Relationen zu qual



\section{Material and Methods}

% 1. Generell erklären, dass fast alles per optionen kontrolliert werden kann.
% 2. 

%What we did
%
%How was the problem studied?
%    Include detailed information, so that a knowledgeable
%         reader can reproduce the experiment
%    Extensive detail can be described in the supplementary
%         material


%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%% BEGIN FIGURE 2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\begin{figure}[t!h]
%\centering \includegraphics[width=0.8\columnwidth]{Overview.jpg}
%\caption{Method...}
%\label{fig:Ceramide}
%\end{figure}
%%%%%%%%%%%%%%%%%%%%%%%%%%%% END FIGURE 2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%% BEGIN FIGURE 2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\begin{figure}[t!h]
%\centering \includegraphics[width=0.8\columnwidth]{DataStructure.jpg}
%\caption{SNP data structure}
%\label{fig:Ceramide}
%\end{figure}
%%%%%%%%%%%%%%%%%%%%%%%%%%%% END FIGURE 2 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\section{Results}
%What have we found out?
%
%Only representative results, essential for the Discussion,
%    should be presented
%Add supplementary materials for data of secondary
%    importance
%Do not attempt to "hide" data in the hope of saving it for a
%    later paper.





\section{Discussion}
%(Short summary: What did you find out?)
%
%What did you learn from your study?
%
%What inferences can be drawn from the findings?
%
%What are the practical and theoretical implications of your
%        findings?
%
%How do your results relate previous research?
%
% (Which problems are next to be solved?)
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%


\section{Conclusion}
%How the work advances the filed from the present state of
%     knowledge
%
%Without clear Conclusions, reviewers and readers will find it
%     difficult to judge the work, and whether or not it merits
%     publication in the journal
%
%Do not repeat the abstract or just list experimental
%     results!!!



\section*{Acknowledgements}
\paragraph{Funding:}
Federal Ministry of Education and Research (BMBF, Germany) in the National
Genome Research Network (NGFN+) under grant number 01GS08134.
\paragraph{Conflict of interest:} None declared.

%\bibliographystyle{natbib}
\bibliographystyle{splncs}

\bibliography{KEGGtranslator_v2_methods}

\end{document} 